reports/helpers.py

```diff
@@ -0,0 +1,252 @@
+                    'billing_end': 30 - 9 - 2018,
+                    'days': 30,
+                    'charge': 300,
+                }
+    """
+    days = 0
+    charge = 0
+    billing_start, billing_end = get_billing_dates(appserver, invoice_month)
+
+    if billing_start and billing_end:
+        # Adding one to the days to include the first day
+        days = billing_end.day - billing_start.day + 1
+        charge = settings.BILLING_RATE * days
+
+    return {
+        'name': appserver.name,
+        'billing_start': billing_start,
+        'billing_end': billing_end,
+        'days': days,
+        'charge': charge,
+    }
+
+
+def get_instance_charges(instance, invoice_month):
+    """
+    Given an instance, this function will be responsible of generating
+    all the charges for the given instance's AppServers
+    :param instance: An openEdxInstance object
+    :param invoice_month: The beginning of the month we want to issue usages for.
+    :return: A pair of the AppServers total usage and a list of the AppServers
+             usages details. Example:
+                [
+                    {
+                        'name': 'AppServer 1',
+                        'billing_start': 1 - 9 - 2018,
+                        'billing_end': 30 - 9 - 2018,
+                        'days': 30,
+                        'charge': 300,
+                    },
+                    {
+                        'name': 'AppServer 2',
+                        'billing_start': 1 - 9 - 2018,
+                        'billing_end': 10 - 9 - 2018,
+                        'days': 10,
+                        'charge': 200,
+                    }
+                ]
+    """
+    appservers_charges = []
+    appservers_total = 0
+
+    if invoice_month > timezone.now():
+        return appservers_charges, appservers_total
+
+    for server in instance.appserver_set.all():
+        charge_details = generate_charge_details(server, invoice_month)
+
+        appservers_charges.append(charge_details)
+        appservers_total += charge_details['charge']
+
+    return appservers_charges, appservers_total
+
+
+def generate_watched_forks_instances(organization):
+    """
+    Given an organization obj, this function will extract all of the watched
+    forks for this specific organization overtime and the instances generated
+    for these watched forks.
+    :param organization: Organization object
+    :return: A dictionary of the watched forks and their instances. Example:
+            {
+                'fork1': [instance 1, instance 2],
+                'fork2': [instance 4, instance 7, instance 6],
+            }
+    """
+    watched_forks = defaultdict(list)
+    pull_requests = WatchedPullRequest.objects.filter(
+        watched_fork__organization=organization
+    ).select_related('watched_fork', 'instance')
+
+    for pull_request in pull_requests:
+        watched_forks[pull_request.watched_fork].append(pull_request.instance)
```

Comment:
In the edge case when a WatchedPullRequest doesn't have an instance (i.e. its instance is None), this will add None to the list. It shouldn't; you can add an `if pull_request.instance:`.
(This „edge case“ happens in Ocim stage and production; we have around 15 WPR like that.)